ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Protocol/Abstract.hs

```diff
@@ -170,13 +176,66 @@ class ( Show (ChainDepState   p)
   -- | We require that protocols support a @k@ security parameter
   protocolSecurityParam :: ConsensusConfig p -> SecurityParam
 
--- | Compare a candidate chain to our own
+-- | The chain order of some type; in the Consensus layer, this will always be
+-- the 'SelectView' of some 'ConsensusProtocol'.
 --
--- If both chains are equally preferable, the Ouroboros class of consensus
--- protocols /always/ sticks with the current chain.
-preferCandidate :: ConsensusProtocol p
-                => proxy      p
-                -> SelectView p  -- ^ Tip of our chain
-                -> SelectView p  -- ^ Tip of the candidate
-                -> Bool
-preferCandidate _ ours cand = cand > ours
+-- See 'preferCandidate' for the primary documentation.
+--
+-- Additionally, we require a total order on this type, such that eg different
```

Comment:
I think we need to reword and/or elaborate this comment.

- My basic concern: it says eg that the total order is used to prioritize candidates in the BlockFetch decision logic. That sure sounds like we're not restricting the VRF tiebreaker in the BlockFetch decision logic.
- I guess because we're not? ... Are there in-flight limits etc that could mean this unrestricted VRF determines which blocks are even fetched?
- Also I think it'd be helpful if this comment explained that the thing these situations have in common is that these decisions are not made in a context where "block arrival order" is not inherently available. (Initial selection and out-of-order blocks would require us to record timestamps in the VolDB. In BlockFetch decision logic, I suppose header arrival order sort of does apply (if you're the only better header, you'll be fetched), unless _multiple_ tied headers arrived between the two BlockFetch invocations----again unless we recorded timestamps.)

Reply:
> * My basic concern: it says eg that the total order is used to prioritize candidates in the BlockFetch decision logic. That sure sounds like we're not restricting the VRF tiebreaker in the BlockFetch decision logic.
> * I guess because we're not? ... Are there in-flight limits etc that could mean this unrestricted VRF determines which blocks are even fetched?

Yes, this could be the case, it is the same as today. Note that we still only download blocks corresponding to chains that are better than the current selection according to `preferCandidate`, so we won't download a chain that is better due to `Ord`, but not due to `preferCandidate`.

> Also I think it'd be helpful if this comment explained that the thing these situations have in common is that these decisions are not made in a context where "block arrival order" is not inherently available. (Initial selection and out-of-order blocks would require us to record timestamps in the VolDB. In BlockFetch decision logic, I suppose header arrival order sort of does apply (if you're the only better header, you'll be fetched), unless _multiple_ tied headers arrived between the two BlockFetch invocations----again unless we recorded timestamps.)

I like the absence of arrival times as a connecting reason for the applications of the `Ord` instance. I am a *bit* hesitant as I am not yet completely sure how the logic would/should look like even if we recorded arrival times there. Eg write `a ⊏ b` for `preferCandidate cfg a b`, and suppose that the BlockFetch decision logic just became aware of the new headers `a,b,c`, which were received in this order, each extending the current selection, with `sv(a) ⊏ sv(b)`, `sv(b) ⊏ sv(c)` and `sv(c) = sv(a)` (`sv` = `selectView`).

 - Downloading `block(a)` first seems weird as `b` is strictly preferrable.
 - Downloading `block(b)` first seems weird as `c` is strictly preferrable.
 - Downloading `block(c)` first seems weird as `a` is just as preferrable as `c`, but it arrived earlier.

Maybe I could just mention "absence of arrival times" as an "intuition", but stress that there are actually no plans to implement arrival times for these use cases?

---

Also FTR, for the particular case of BlockFetch, [this paper ("the Tse paper")](https://eprint.iacr.org/2021/1545.pdf) that we might want prioritize downloading blocks with higher slot numbers, not block numbers. The core issue remains though for initial chain selection and out-of-order arrival.

Reply:
> Note that we still only download blocks corresponding to chains that are better than the current selection according to preferCandidate, so we won't download a chain that is better due to Ord, but not due to preferCandidate.

Good point 👍. So the "risk" is the same as in ChainSel.

> I am a bit hesitant as I am not yet completely sure how the logic would/should look like even if we recorded arrival times there

Great point! And thanks for the concrete example. I do think I see a clear path, though: treat it the exact same as ChainSel. In particular: fetch whichever of `a`, `b`, and `c` are better than _our current selection_ (regardless of how they compare to each other[^1]). Fetch them in their arrival order, since in-flight limits might not allow us to fetch them all immediately.

That resolves it, yeah?

[^1]: Although maybe I'm assuming they're all the same length? Don't download _shorter_ ones (yet). But I think that is veering towards a topo sort.

Reply:
So you are suggesting to completely ignore tiebreakers to prioritize between different chains that are preferred over our selection? I.e. sorting on `(BlockNo, Down ArrivalTime)`? That makes sense (the Tse rule would be another option); I now changed this paragraph by adding a sentence in the end:

```haskell
-- Additionally, we require a total order on this type, such that eg different
-- candidate chains that are preferred over our current selection can be sorted
-- for prioritization. For example, this is used in ChainSel during initial
-- chain selection or when blocks arrive out of order (not the case when the
-- node is caught up), or in the BlockFetch decision logic. Future work could
-- include also recording\/storing arrival information and use that instead
-- of\/in addition to the 'Ord' instance.
```

Reply:
Yeah, I think that captures it well.